File: typescript/src/models/Meta.ts

export interface MetaProps {
    comments?: string[];
    attr?: Record<string, any>;
    tags?: string[];
}

/**
 * Mixin/Base class that adds metadata storage capabilities to Node and Schema.
 */
export class Meta {
    comments: string[];
    attr: Record<string, any>;
    tags: string[];

    constructor({ comments, attr, tags }: MetaProps = {}) {
        this.comments = comments ? [...comments] : [];
        this.attr = attr ? { ...attr } : {};
        this.tags = tags ? [...tags] : [];
    }

    /**
     * Clears ALL metadata (comments, attributes, tags).
     */
    clearCommonMeta(): void {
        this.comments = [];
        this.attr = {};
        this.tags = [];
    }

    /**
     * Merges only the common fields (attributes, comments, tags)
     * from a MetaInfo object. Safe for both Node and Schema.
     */
    applyCommonMeta(info: Meta): void {
        // Append comments
        if (info.comments.length > 0) {
            this.comments.push(...info.comments);
        }

        // Merge attributes ($key=value)
        if (Object.keys(info.attr).length > 0) {
            Object.assign(this.attr, info.attr);
        }

        // Append tags
        if (info.tags.length > 0) {
            this.tags.push(...info.tags);
        }
    }
}

/**
 * A temporary container (DTO) holding parsed metadata from a / ... / block.
 * It contains BOTH Schema constraints (!required) and Node attributes ($key=val).
 */
export class MetaInfo extends Meta {
    required: boolean;

    constructor(props: MetaProps & { required?: boolean } = {}) {
        super(props);
        this.required = props.required || false;
    }

    /**
     * Merges everything from another MetaInfo object, including 'required' flag.
     */
    applyMeta(info: MetaInfo): void {
        // Apply common fields (comments, attr, tags)
        this.applyCommonMeta(info);

        // Override required meta (Schema Only)
        if (info.required) {
            this.required = true;
        }
    }

    /**
     * Allows usage checks like 'if meta.isEmpty()' to see if any metadata was collected.
     * Equivalent to Python's __bool__.
     */
    isEmpty(): boolean {
        return (
            this.comments.length === 0 &&
            Object.keys(this.attr).length === 0 &&
            this.tags.length === 0 &&
            !this.required
        );
    }

    /**
     * Debug representation mimicking the actual ADF format style.
     * Example: <MetaInfo !required #tag $key=val >
     */
    toString(): string {
        const parts: string[] = [];

        // 1. Flags
        if (this.required) {
            parts.push("!required");
        }

        // 2. Tags
        for (const t of this.tags) {
            parts.push(`#${t}`);
        }

        // 3. Attributes
        for (const [k, v] of Object.entries(this.attr)) {
            // Simplistic value repr for debug
            let valStr: string;
            if (typeof v === 'string') {
                valStr = `"${v}"`;
            } else {
                valStr = String(v);
            }
            parts.push(`$${k}=${valStr}`);
        }

        // 4. Comments (Summary)
        if (this.comments.length > 0) {
            if (this.comments.length === 1) {
                const c = this.comments[0];
                const preview = c.length > 15 ? c.substring(0, 15) + '..' : c;
                parts.push(`/* ${preview} */`);
            } else {
                parts.push(`/* ${this.comments.length} comments */`);
            }
        }

        const content = parts.join(" ");
        return content ? `<MetaInfo ${content}>` : "<MetaInfo (empty)>";
    }

}

File: python/src/arkadia/ai/data/Meta.py
from typing import Optional, List, Dict, Union, Any
from enum import Enum


class MetaInfo:
    """
    A temporary container (DTO) holding parsed metadata from a / ... / block.
    It contains BOTH Schema constraints (!required) and Node attributes ($key=val).
    """
    def __init__(self):
        # Common (Node & Schema)
        self.attr: Dict[str, Any] = {}        # $key=value
        self.comments: List[str] = []         # /* comments */
        self.tags: List[str] = []             # #["tag1"]

        # Schema Only (Constraints)
        self.required: bool = False           # !required

    def apply_meta(self, info: "MetaInfo"):
        # Append comments
        if info.comments:
            self.comments.extend(info.comments)

        # Merge attributes ($key=value)
        if info.attr:
            self.attr.update(info.attr)
            
        # Append tags
        if info.tags:
            self.tags.extend(info.tags)

        # Override required meta
        self.required = info.required

    def __bool__(self):
        """
        Allows usage: 'if meta:' to check if any metadata was collected.
        Returns False if the container is effectively empty.
        """
        return bool(self.attr or self.tags or self.comments or self.required)
    

    def __repr__(self):
        """
        Debug representation mimicking the actual ADF format style.
        Example: <MetaInfo !required #tag $key=val /* 2 comments */>
        """
        parts = []

        # 1. Flags
        if self.required:
            parts.append("!required")

        # 2. Tags
        for t in self.tags:
            parts.append(f"#{t}")

        # 3. Attributes
        for k, v in self.attr.items():
            # Simplistic value repr
            val_str = str(v).lower() if isinstance(v, bool) else repr(v)
            parts.append(f"${k}={val_str}")

        # 4. Comments (Summary)
        if self.comments:
            if len(self.comments) == 1:
                # Show short comment content if singular
                c = self.comments[0]
                preview = (c[:15] + '..') if len(c) > 15 else c
                parts.append(f"/* {preview} */")
            else:
                parts.append(f"/* {len(self.comments)} comments */")

        content = " ".join(parts)
        return f"<MetaInfo {content}>" if content else "<MetaInfo (empty)>"

class Meta:
    """
    Mixin class that adds metadata storage capabilities to Node and Schema.
    """
    def __init__(self, 
                 comments: Optional[List[str]] = None,
                 attr: Optional[Dict[str, Any]] = None, 
                 tags: Optional[List[str]] = None):
        self.comments = comments if comments is not None else []
        self.attr = attr if attr is not None else {}
        self.tags = tags if tags is not None else []

    
    def clear_common_meta(self):
        """
        Clears ALL metadata,
        """
        self.comments = []
        self.attr = {}
        self.tags = []


    def apply_common_meta(self, info: 'MetaInfo'):
        """
        Merges only the common fields (attributes, comments, tags) 
        from a MetaInfo object. Safe for both Node and Schema.
        """

        # Append comments
        if info.comments:
            self.comments.extend(info.comments)

        # Merge attributes ($key=value)
        if info.attr:
            self.attr.update(info.attr)
            
        # Append tags
        if info.tags:
            self.tags.extend(info.tags)